feat: Added missing info to IAP transaction and product structures

kevinlinv opened the pull request:
#### Description of Change

The In-app purchase is a module that allows us to communicate with Apple's Mac App Store by using Store Kit. It allows us to get product information to be displayed to users and make purchases if users choose to do so.

There are a few missing fields in the data structures used for in-app purchase, specifically subscription information on `Product` and payment discount information on `Transaction`. Referencable from:
- `Product`: https://developer.apple.com/documentation/storekit/skproduct
- `Transaction`: https://developer.apple.com/documentation/storekit/skpaymenttransaction
  - `payment`: https://developer.apple.com/documentation/storekit/skpayment

This PR register those missing fields to the structure definition, get and convert the data to be returned.

Resolves https://github.com/electron/electron/issues/31738. 

###  ⚠️  Disclaimer ⚠️

This is the first time I've ever coded in C++ or Objective-C I mostly followed prior PR: https://github.com/electron/electron/pull/25058. I would be happy for maintainers to take it over if it would get the PR move along faster.

~~Also, I wasn't able to get electron build locally (running to errors when using build-tools), is there any way for me to get a build of electron for mac from CI so that I could test the in-app purchase?~~ The artifacts seem to be uploaded in CI, given that I can pass the build that is.

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included and stakeholders cc'd
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: feat: Added missing info to in-app purchase's `transaction` and `product` structures

kevinlinv commented:
Would love to get this backported to v15 if possible.

codebytere commented:
hey @kevinlinv! thanks for this - we'll start the review process as soon as we can!

kevinlinv commented:
👋  Hey, thanks for the heads up.

I managed to produce a build on CI, I did some testing and found that some of the data are not set to `undefined` when they should have. Going to make some code changes and see if I can fix that today otherwise might need to get some help from reviewers later 👌

kevinlinv commented on shell/browser/mac/in_app_purchase_observer.mm:
```diff
@@ -94,6 +94,37 @@ - (NSString*)dateToISOString:(NSDate*)date {
+        [paymentDiscount.keyIdentifier UTF8String];
+  }
+
+  // if (paymentDiscount.nonce != nil) {
+  //   paymentDiscountStruct.nonce = [paymentDiscount.nonce uuidString];
+  // }
```
I am not quite sure how to get the UUID value here. 
Appreciate if I get some help here 🙏

kevinlinv replied:
From [Apple docs](https://developer.apple.com/documentation/storekit/skpaymentdiscount/3043531-nonce), it is typed as UUID:

![image](https://user-images.githubusercontent.com/44967897/149259292-d2cb1162-c7d7-4064-a471-0ee9e633d691.png)

I'd imagine that we would expose it in JS in `string` type.

I wanted to figure out how I could pull the Objective C UUID and expose it as `string` in JS.
I got a syntax error trying to read the UUID value with `[paymentDiscount.nonce uuidString]`

kevinlinv commented on shell/browser/mac/in_app_purchase_product.mm:
```diff
@@ -156,9 +216,64 @@ - (NSString*)formatPrice:(NSDecimalNumber*)price
 
   // Downloadable Content Information
   productStruct.isDownloadable = [product downloadable];
+  if (@available(macOS 10.14, *)) {
+    if (product.downloadContentVersion != nil) {
+      productStruct.downloadContentVersion =
+          [product.downloadContentVersion UTF8String];
+    }
+    if (product.downloadContentLengths != nil) {
+      productStruct.downloadContentLengths.reserve(
+          [product.downloadContentLengths count]);
+
+      for (NSNumber* contentLength in product.downloadContentLengths) {
+        productStruct.downloadContentLengths.push_back(
+            [contentLength longLongValue]);
+      }
+    }
+  } else {
+    if (product.contentVersion != nil) {
+      productStruct.downloadContentVersion =
+          [product.contentVersion UTF8String];
+    }
+    if (product.contentLengths != nil) {
+      productStruct.downloadContentLengths.reserve(
+          [product.contentLengths count]);
+
+      for (NSNumber* contentLength in product.contentLengths) {
+        productStruct.downloadContentLengths.push_back(
+            [contentLength longLongValue]);
+      }
+    }
+  }
```
`downloadContentVersion` and `downloadContentLengths` are intended to replace `contentVersion` and `contentLengths` respectively. http://codeworkshop.net/objc-diff/sdkdiffs/macos/10.14/StoreKit.html

kevinlinv commented on shell/browser/api/electron_api_in_app_purchase.cc:
```diff
@@ -23,6 +39,12 @@ struct Converter<in_app_purchase::Payment> {
     dict.SetHidden("simple", true);
     dict.Set("productIdentifier", payment.productIdentifier);
     dict.Set("quantity", payment.quantity);
+    if (!payment.applicationUsername.empty()) {
+      dict.Set("applicationUsername", payment.applicationUsername);
+    }
+    if (!payment.paymentDiscount.identifier.empty()) {
+      dict.Set("paymentDiscount", payment.paymentDiscount);
+    }
```
Not quite sure whether this is the right way to set an `undefined` field.
I tried setting it in Objective C code, but it doesn't seem to work.

kevinlinv replied:
Hmmm, ok, it looks like the other fields that belong here also do the same thing. 
I have removed all optional for now to make them consistent.

kevinlinv commented:
👋  a quick ping here!

I am really keen to get this PR going as soon as possible, my app offers subscriptions and has promotional offers that uses `product.introductoryPrice` in some cases.

kevinlinv commented:
cc @codebytere 

I just read the recent announcement from the electron team about an incoming break for the entirety of December in this blog post https://www.electronjs.org/blog/a-quiet-place-21.

I am hoping that we can land this before then. Would be super if we can get someone assigned as a reviewer 👍

kevinlinv commented:
Heya @codebytere @jkleinsc and the rest of electron team, happy new year 🎉 !

I am very keen on this PR to go through to avoid having to create separate node modules for IAP. 
Pinging here to get a review on this 🏓

jkleinsc reviewed (commented):
API LGTM

kevinlinv commented:
@jkleinsc thanks for the API review.

I really want this to be pushed through. Are you able to suggest someone that can review this PR? 
It would be better if we can get someone assigned too. 🙏

kevinlinv commented:
@codebytere @jkleinsc 

Apologies for the frequent ping 🏓 

The PR has been open for quite some time (_it's been more or less a month discounting the electron team break_) and I am really keen to get this moving 🙂  Would you be able to review (if not, suggest/assign a reviewer for) this pull request, please?

nornagon commented on shell/browser/mac/in_app_purchase_product.h:
```diff
@@ -27,12 +50,16 @@ struct Product {
+  ProductDiscount introductoryPrice;
+  std::vector<ProductDiscount> discounts;
+  std::string subscriptionGroupIdentifier;
+  ProductSubscriptionPeriod subscriptionPeriod;
```
`absl::optional` here too I think?

kevinlinv commented on shell/browser/mac/in_app_purchase_observer.h:
```diff
@@ -21,9 +21,26 @@ namespace in_app_purchase {
   std::string productIdentifier = "";
   int quantity = 1;
+  std::string applicationUsername;
+  PaymentDiscount paymentDiscount;
```
I have removed the optionality for now to follow suit with how other optional values are swapped with empty value.

kevinlinv commented on shell/browser/api/electron_api_in_app_purchase.cc:
```diff
@@ -15,6 +15,22 @@
+    dict.SetHidden("simple", true);
+    dict.Set("identifier", paymentDiscount.identifier);
+    dict.Set("keyIdentifier", paymentDiscount.keyIdentifier);
+    // dict.Set("nonce", paymentDiscount.nonce);
```
Yeah, I am not sure how to get the value out from the Objective C code.

kevinlinv commented on shell/browser/mac/in_app_purchase_observer.mm:
```diff
@@ -94,6 +94,37 @@ - (NSString*)dateToISOString:(NSDate*)date {
+
+  if (paymentDiscount.identifier != nil) {
+    paymentDiscountStruct.identifier = [paymentDiscount.identifier UTF8String];
+  }
```
I am getting an error after removing the `nil` check originating from `skPaymentToStruct` which I believe was thrown because I removed the `nil` check for `paymentStruct.applicationUsername = [payment.applicationUsername UTF8String];` assignment 🤔 

I've re-added the `nil` check back to properties that are optional based on apple's doc.

```
../../buildtools/third_party/libc++/trunk/include/string:2551: _LIBCPP_ASSERT '__s != nullptr' failed. string::assign received nullptr
Received signal 6
0   Electron Framework                  0x0000000111ad6529 base::debug::CollectStackTrace(void**, unsigned long) + 9
1   Electron Framework                  0x00000001119fc243 base::debug::StackTrace::StackTrace() + 19
2   Electron Framework                  0x0000000111ad6481 base::debug::(anonymous namespace)::StackDumpSignalHandler(int, __siginfo*, void*) + 1409
3   libsystem_platform.dylib            0x00007fff204d7d7d _sigtramp + 29
4   libdyld.dylib                       0x00007fff204acce8 _dyld_fast_stub_entry(void*, long) + 65
5   libsystem_c.dylib                   0x00007fff203e7406 abort + 125
6   Electron Framework                  0x0000000116e18132 std::__1::__libcpp_abort_debug_function(std::__1::__libcpp_debug_info const&) + 82
7   Electron Framework                  0x000000010d6764cf std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::operator=(char const*) + 79
8   Electron Framework                  0x000000010d6766b0 -[InAppTransactionObserver skPaymentToStruct:] + 240
9   Electron Framework                  0x000000010d676ab0 -[InAppTransactionObserver skPaymentTransactionToStruct:] + 736
10  Electron Framework                  0x000000010d675fe5 -[InAppTransactionObserver runCallback:] + 341
11  StoreKit                            0x00007fff5403e617 __48-[SKPaymentQueue _updatedTransactions:restored:]_block_invoke_2 + 62
12  libdispatch.dylib                   0x00007fff202e7623 _dispatch_call_block_and_release + 12
13  libdispatch.dylib                   0x00007fff202e8806 _dispatch_client_callout + 8
14  libdispatch.dylib                   0x00007fff202eae37 _dispatch_queue_override_invoke + 775
15  libdispatch.dylib                   0x00007fff202f7818 _dispatch_root_queue_drain + 326
16  libdispatch.dylib                   0x00007fff202f7f70 _dispatch_worker_thread2 + 92
17  libsystem_pthread.dylib             0x00007fff2048f417 _pthread_wqthread + 244
18  libsystem_pthread.dylib             0x00007fff2048e42f start_wqthread + 15
[end of stack trace]
```

kevinlinv commented:
@nornagon thank you for the review! 🙌 

Quick summary of changes applied after your feedback:
1. On exposing empty value to JS when there one from Objective C is unset — I've changed all to do as you suggested, it also follows the existing pattern of the module 👍 
2. On why `nonce` is commented out — I've left a comment.
3. On removing `nil` check — I found out that removing those checks actually causes the app to crash for some reason

PTAL again 🙂

nornagon commented on shell/browser/mac/in_app_purchase_observer.mm:
```diff
@@ -94,6 +94,25 @@ - (NSString*)dateToISOString:(NSDate*)date {
+  paymentDiscountStruct.identifier = [paymentDiscount.identifier UTF8String];
+  paymentDiscountStruct.keyIdentifier =
+      [paymentDiscount.keyIdentifier UTF8String];
+  // paymentDiscountStruct.nonce = [paymentDiscount.nonce uuidString];
```
try `-[NSUUID UUIDString]`: https://developer.apple.com/documentation/foundation/nsuuid/1416585-uuidstring?language=objc
```suggestion
  paymentDiscountStruct.nonce = [paymentDiscount.nonce UUIDString];
```

kevinlinv replied:
Just pushed the commit 👍

nornagon commented on docs/api/structures/product.md:
```diff
@@ -8,4 +8,11 @@
+available for the product.
+* `discounts` [ProductDiscount](product-discount.md)[] - An array of discount offers
+* `subscriptionGroupIdentifier` string - The identifier of the subscription group to which the subscription belongs.
+* `subscriptionPeriod` [ProductSubscriptionPeriod](product-subscription-period.md) - The period details for products that are subscriptions.
```
& same for `ProductSubscriptionPeriod | undefined`.

An empty string is one thing—it can be checked with `if (foo.contentVersion) { /* ... */ }`—but an empty object feels off. Arguably the empty strings aren't a great way to express "not present" either, but it's at least consistent with the existing API.

kevinlinv replied:
Ah, alright, I'll check the typing again and make objects optional if data within it isn't set.

kevinlinv commented on docs/api/structures/product.md:
```diff
@@ -8,4 +8,11 @@
 * `price` number - The cost of the product in the local currency.
 * `formattedPrice` string - The locale formatted price of the product.
 * `currencyCode` string - 3 character code presenting a product's currency based on the ISO 4217 standard.
+* `introductoryPrice` [ProductDiscount](product-discount.md) - The object containing introductory price information for the product.
+available for the product.
```
Ah, alright, I'll check the typing again and make objects optional if data within it isn't set.

kevinlinv commented:
@nornagon thank you for the quick feedback loop.

Apologies for the slowness, I am do not have the setup locally and not well versed with the syntax 😅 
I've addressed your feedback, PTAL again 👍

kevinlinv commented:
🏓  @nornagon

nornagon commented:
@kevinlinv please don't ping excessively. I have a lot of issues on my plate, and I will get to this one when I have the time to do so.

nornagon commented on shell/browser/mac/in_app_purchase_product.h:
```diff
@@ -14,6 +14,29 @@ namespace in_app_purchase {
+  std::string priceLocale;
+  int paymentMode;
+  int numberOfPeriods;
+  ProductSubscriptionPeriod subscriptionPeriod;
```
This should be optional, right?

nornagon replied:
Hm, it returns a pointer, can that pointer be `nil`? that would imply that the subscription period is optional.

kevinlinv replied:
Ah, yeah, that's true, I'll convert this to optional then 👌

kevinlinv replied:
Done! 👍

nornagon commented on docs/api/structures/product-discount.md:
```diff
@@ -0,0 +1,9 @@
+* `type` number - The type of discount offer.
+* `price` number - The discount price of the product in the local currency.
+* `priceLocale` string - The locale used to format the discount price of the product.
+* `paymentMode` number - The payment mode for this product discount. `pay_as_you_go: 0`, `pay_up_front: 1`, `free_trial: 2`.
```
API-wise, I think exposing this as a string is better than a number. This should be
```suggestion
* `paymentMode` string - The payment mode for this product discount. Can be `payAsYouGo`, `payUpFront`, or `freeTrial`.
```

nornagon commented on docs/api/structures/product-subscription-period.md:
```diff
@@ -0,0 +1,4 @@
+# ProductSubscriptionPeriod Object
+
+* `numberOfUnits` number - The number of units per subscription period.
+* `unit` number - The increment of time that a subscription period is specified in. `day: 0`, `week: 1`, `month: 2`, `year: 3`.
```
Same here.

nornagon commented on shell/browser/mac/in_app_purchase_product.h:
```diff
@@ -9,11 +9,35 @@
+  int type;
+  double price = 0.0;
+  std::string priceLocale;
+  int paymentMode;
```
```suggestion
  SKProductDiscountPaymentMode paymentMode;
```

nornagon commented on shell/browser/mac/in_app_purchase_product.mm:
```diff
@@ -106,6 +106,61 @@ - (NSString*)formatPrice:(NSDecimalNumber*)price
+    (SKProductDiscount*)productDiscount API_AVAILABLE(macosx(10.13.2)) {
+  in_app_purchase::ProductDiscount productDiscountStruct;
+
+  productDiscountStruct.paymentMode = (int)productDiscount.paymentMode;
```
```suggestion
  productDiscountStruct.paymentMode = productDiscount.paymentMode;
```

nornagon commented on shell/browser/api/electron_api_in_app_purchase.cc:
```diff
@@ -45,6 +65,38 @@ struct Converter<in_app_purchase::Transaction> {
+    dict.Set("type", productDiscount.type);
+    dict.Set("price", productDiscount.price);
+    dict.Set("priceLocale", productDiscount.priceLocale);
+    dict.Set("paymentMode", productDiscount.paymentMode);
```
```suggestion
    if (productDiscount.paymentMode == SKProductDiscountPaymentModePayAsYouGo) {
      dict.Set("paymentMode", "payAsYouGo");
    } else if (/* ... */) { /* ... */ }
```

You might need to fiddle around with the enum, not sure if you can #include the relevant header here or not.

kevinlinv replied:
Nice, thanks for the suggestion.
I tried to do it in the C++ land, but the import becomes a bit complicated and I am not sure how to do it properly.
I decided to do the enum switching in Objective-C instead, where the typing has already properly imported 👍

nornagon commented on shell/browser/mac/in_app_purchase_product.h:
```diff
@@ -9,11 +9,35 @@
 
+struct ProductSubscriptionPeriod {
+  int numberOfUnits;
+  int unit;
```
```suggestion
  SKProductPeriodUnit unit;
```

kevinlinv commented:
Thanks for the review 🎉  
I shall rebase this PR and then leave it for the team to merge 🙂

nornagon commented:
Changing backport labels because we don't backport new features to stable branches without @electron/wg-releases approval.
